### ISSUE TITLE
Add option to provide S3 object user metadata when writing files to S3 

### ISSUE THREAD

**Issue by Laurens-makel** [ibissource/iaf#4865]
**❗ please do not add sensitive information in issues, you can provide extra information via email using issue number as reference ❗**

**Describe the issue**
The Frank Framework is missing support to provide user metadata when using the S3 sender whilst the `s3Client.putObject()` operation has support to provide user metadata to the S3 object.

In our use case we would like to preserve the original filename in the metadata of the S3 object.

![image](https://user-images.githubusercontent.com/37303445/236455591-08ca8e96-6467-476f-9c9d-a62e6fe0a10f.png)

**Reporter**
Laurens

**Comment by jacodg** [comment]
Can you give us your custom code based on 7.7 that implements this?

**Comment by Laurens-makel** [comment]
Yes, the following code did the trick, it only supports the `metadata-filename` parameter to set a hard coded value for our use case, but with some little modifications it could be modified into a generic solution.

First, I declared an interface named `FileSystemWithMetadata` and adjusted the AmazonS3FileSystem to implement that interface.

```
package nl.nn.adapterframework.filesystem;

import java.util.Map;

public interface FileSystemWithMetadata<F> {
    
    public F toFileWithMetadata(String filename, Map<String, String> metadata) throws FileSystemException;
    
}
```

Read the metadata that needs to be set from pipeline session In FileSystemActor, case WRITE:
![image](https://github.com/ibissource/iaf/assets/37303445/66813302-1a3c-4265-a5f0-0642c99193cc)
```
F file= fileSystem instanceof FileSystemWithMetadata
							? ((FileSystemWithMetadata<F>) fileSystem).toFileWithMetadata(determineFilename(input, pvl), getMetadata(pvl))
							: getFile(input, pvl);
```

```

	private Map<String, String> getMetadata(ParameterValueList pvl) {
		Map<String, String> metadata = new HashMap();

		metadata.put("filename", pvl.get("metadata-filename").asStringValue());

		return metadata;
	}
```


Support the creation of files with metadata by implementing `toFileWithMetadata` in AmazonS3FileSystem.java

```
@Override
public S3Object toFileWithMetadata(String filename, Map<String, String> metadata) throws FileSystemException {
    S3Object object = toFile(filename);
    object.getObjectMetadata().setUserMetadata(metadata);

    return object;
}
```

Copy metadata from `f` to new metadata instance before calling `s3Client.putObject()` to save it in AmazonS3FileSystem.java.

```
ObjectMetadata metaData = new ObjectMetadata();
...
Map<String, String> userMetadata = f.getObjectMetadata().getUserMetadata();
if(userMetadata != null){
    metaData.setUserMetadata(userMetadata);
}
...
s3Client.putObject(bucketName, f.getKey(), fis, metaData)
```

**Comment by Laurens-makel** [comment]
![image](https://github.com/ibissource/iaf/assets/37303445/6d472e19-8749-49a5-9188-ac7df09530cf)

**Comment by Laurens-makel** [comment]
This seems to be the only custom code left in our custom S3 setup left, it would be nice if we can close this issue and remove the custom code.

What do you think about above implementation? If agreed, I could work it out. Else, could you suggest the favoured way to go? I can have a look at creating a PR.
